File: base/scripts/containers/system_tools.lua
require("containers/debug_panel")
require('containers/console_panel')
require('containers/widget_tree')

class "SystemTools" (Container)

function SystemTools:init()
    self:setOrder(1000)
    
    self:addCallback("KeyDown_Grave", self.viewConsole, self)
    self:addCallback("KeyDown_F1",    self.viewSystemInfo, self)
    self:addCallback("KeyDown_F2",    self.viewDebugView, self)
    self:addCallback("KeyDown_F3",    self.viewWidgetsTree, self)
end

function SystemTools:reset()
    if (self.widgets_tree) then
        self.widgets_tree:instantView(false)
    end
end

function SystemTools:viewConsole()
    if (not self.console) then
        self.console = ConsolePanel()
        self:attach(self.console)
    end
    if (self.console) then
        self.console:view(not self.console:isOpened())
    end
end

function SystemTools:viewSystemInfo()
    if (not self.debug_panel) then
        self.debug_panel = DebugPanel()
        self:attach(self.debug_panel)
    end
    if (self.debug_panel) then
        self.debug_panel:view(not self.debug_panel:isOpened())
    end
end

function SystemTools:viewDebugView()
    local current_screen = UserSave:getCurrentScreen()
    if (current_screen) then
        current_screen:switchDebugView()
    end
end

function SystemTools:viewWidgetsTree()
    if (not self.widgets_tree) then
        self.widgets_tree = WidgetTree()
        self:attach(self.widgets_tree)
    end
    if (self.widgets_tree) then
        self.widgets_tree:view(not self.widgets_tree:isOpened())
    end
end

function SystemTools:log(...)
    if (not self.console) then
        self.console = ConsolePanel()
        self:attach(self.console)
    end
    if (self.console) then
        self.console:log(...)
    end
end

File: src/system_tools.cpp
#include "system_tools.h"

#include "debug_panel.h"
#include "console.h"
#include "widgets_tree.h"
namespace redrevolt
{
SystemTools::SystemTools(const std::string & id)
    : Container(id)
    , m_console(nullptr)
    , m_widgetsTree(nullptr)
    , m_debugPanel(nullptr)
{
    addCallback("KeyDown_Grave", this, &SystemTools::viewConsole);
    addCallback("KeyDown_F1", this, &SystemTools::viewSystemInfo);
    addCallback("KeyDown_F2", this, &SystemTools::viewDebugView);
    addCallback("KeyDown_F3", this, &SystemTools::viewWidgetsTree);

    m_console = new Console("console");
    attach(m_console);
    m_widgetsTree = new WidgetsTree("widgetsTree");
    attach(m_widgetsTree);
    m_debugPanel = new DebugPanel("debugPanel");
    attach(m_debugPanel);

    setOrder(999);
}

SystemTools::~SystemTools()
{
    m_console = nullptr;
    m_widgetsTree = nullptr;
    m_debugPanel = nullptr;
}

void SystemTools::viewConsole(stren::Widget * sender)
{
    if (m_console)
    {
        m_console->view(!m_console->isOpened());
    }
}

void SystemTools::viewSystemInfo(stren::Widget * sender)
{
    if (m_debugPanel)
    {
        m_debugPanel->view(!m_debugPanel->isOpened());
    }
}

void SystemTools::viewDebugView(stren::Widget * sender)
{
    stren::Screen * screen = stren::Engine::getCurrentScreen();
    if  (screen)
    {
        screen->setDebugView(!screen->isDebugView());
    }
}

void SystemTools::viewWidgetsTree(stren::Widget * sender)
{
    if (m_widgetsTree)
    {
        m_widgetsTree->view(!m_widgetsTree->isOpened());
    }
}

void SystemTools::log(const char * msg)
{
    if (m_console)
    {
        m_console->log(msg);
    }
}

void SystemTools::reset()
{
    if (m_widgetsTree)
    {
        m_widgetsTree->instantClose();
    }
}
} // redrevolt
